### PR TITLE
Warp e2e

### PR DESCRIPTION
**Pull request by aaronbuchwald**
## Why this should be merged

This PR implements the warp precompile. Addresses https://github.com/ava-labs/subnet-evm/issues/440

## How this works

This PR adds a warp precompile to integrate warp messaging into Subnet-EVM. This includes a Solidity interface for contracts that want to send a message via warp here.

## How this was tested

This code has been tested through precompile unit tests, an e2e VM level test for the VM's handling of verification and acceptance of warp interactions, and an e2e ginkgo test that sends a warp message from subnet A to subnet B where the validator sets of subnet A and B are disjoint.

## How is this documented

https://github.com/ava-labs/subnet-evm/blob/warp-e2e/precompile/contracts/warp/README.md

## TODO

- [x] Split out network runner changes + minor changes to load test into https://github.com/ava-labs/subnet-evm/pull/579
- [x] Move cleanup into a separate PR: (Receipt handling fix: https://github.com/ava-labs/subnet-evm/pull/586#discussion_r1144821965, BLS key setup for unit tests: https://github.com/ava-labs/subnet-evm/pull/586#discussion_r1144822689)
- [x] Clean up parsing of AddressedPayload within an UnsignedMessage
- [x] Improve helpers for unpacking predicate bytes
- [x] Clean up comments and add README to warp precompile package
- [x] Add pre-verification gas charges to `IntrinsicGas` to account for work done in predicate verification
- [x] Add support for receiving messages from Primary Network
- [x] Add synchronous signature aggregator API call
- [x] Benchmark precompile including parsing, P-Chain validator set lookup, and signature verification to calculate gas costs
- [x] Add unit test for VerifyWithContext on an already validated block with a new context that is valid/invalid
- [x] Add warp precompile contract execution unit tests
- [x] Add warp precompile predicate unit tests
- [ ] Stress testing
- [ ] Cleanup and review

### CODE REVIEW

File: contract-examples/test/warp.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+var task = process.env.WARP_TEST_TASK || "send"
+
+describe("Warp", function () {
+  before(async function () {
+    fundedSigner = await ethers.getSigner(fundedAddress)
+    contract = await ethers.getContractAt("WarpMessenger", WARP_ADDRESS, fundedSigner)
+    console.log(`Precompile contract handle at address: ${contract.address}`)
+  })
+
+  it("warp action", async function () {
```
Comment (aaronbuchwald) [inline contract-examples/test/warp.ts:30]: TODO: implement HardHat test for verification of expected contents

File: plugin/evm/vm_warp_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	require.NoError(t, block2VerifyWithCtx.VerifyWithContext(context.Background(), validProposerCtx))
+	require.Equal(t, choices.Processing, block2.Status())
+	require.NoError(t, vm.SetPreference(context.Background(), block2.ID()))
+	require.NoError(t, block2.Accept(context.Background()))
+
+	ethBlock := block2.(*chain.BlockWrapper).Block.(*Block).ethBlock
+	verifiedMessageReceipts := vm.blockChain.GetReceiptsByHash(ethBlock.Hash())
+	require.Len(t, verifiedMessageReceipts, 1)
+	verifiedMessageTxReceipt := verifiedMessageReceipts[0]
+	require.Equal(t, types.ReceiptStatusSuccessful, verifiedMessageTxReceipt.Status)
```
Comment (aaronbuchwald) [inline plugin/evm/vm_warp_test.go:226]: TODO: add more thorough testing across different valid/invalid signatures and using `BuildWithContext/VerifyWithContext`

File: tests/utils/runner/runner.go [hunk 3]
```diff
@@ -73,30 +75,28 @@ func NewDefaultANRConfig() ANRConfig {
 
 // NewNetworkManager constructs a new instance of a network manager
 func NewNetworkManager(config ANRConfig) *NetworkManager {
-	return &NetworkManager{
+	manager := &NetworkManager{
```
Comment (aaronbuchwald) [inline tests/utils/runner/runner.go:78]: TODO: move to separate PR

File: plugin/evm/block.go [hunk 4]
```diff
@@ -98,8 +98,8 @@ func (b *Block) handlePrecompileAccept(sharedMemoryWriter *sharedMemoryWriter) e
 
 	// Read the receipts
 	receipts := rawdb.ReadReceipts(b.vm.chaindb, b.ethBlock.Hash(), b.ethBlock.NumberU64(), b.vm.chainConfig)
-	if receipts == nil {
-		return fmt.Errorf("failed to read receipts for accepted block %s, height %d", b.ethBlock.Hash(), b.ethBlock.NumberU64())
+	if len(receipts) == 0 && b.ethBlock.ReceiptHash() != types.EmptyRootHash {
```
Comment (aaronbuchwald) [inline plugin/evm/block.go:101]: TODO: add a comment explaining this change. This can be moved to a separate PR

File: plugin/evm/vm_test.go [hunk 5]
```diff
@@ -141,6 +144,12 @@ func NewContext() *snow.Context {
 			return subnetID, nil
 		},
 	}
+	blsSecretKey, err := bls.NewSecretKey()
```
Comment (aaronbuchwald) [inline plugin/evm/vm_test.go:147]: This can be moved to a separate PR

File: plugin/evm/vm_warp_test.go (hunk 2 above)
Comment (aaronbuchwald) [reply plugin/evm/vm_warp_test.go:226]: We should test at least the following cases:
- non-zero number of signatures, but does not meet the threshold
- 100% of validators sign
- no signatures
- outdated signatures (later P-Chain height where validator set has changed)
- `VerifyWithContext` on an already verified block with a new valid/invalid context for the predicate

File: precompile/contracts/warp/config.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	// typecast before comparison
+	other, ok := (s).(*Config)
+	if !ok {
+		return false
+	}
+	equals := c.Upgrade.Equal(&other.Upgrade)
+	return equals && c.QuorumNumerator == other.QuorumNumerator
+}
+
+func (c *Config) Accept(acceptCtx *precompileconfig.AcceptContext, txHash common.Hash, logIndex int, topics []common.Hash, logData []byte) error {
```
Comment (aaronbuchwald) [inline precompile/contracts/warp/config.go:94]: TODO: add comments

File: precompile/contracts/warp/config.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	_, err = warpMessages.ParseAddressedPayload(warpMsg.UnsignedMessage.Payload)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse warp payload into addressed payload: %w", err)
+	}
+
+	log.Info("verifyingWarpMessage", "warpMsg", warpMsg, "quorumNum", quorumNumerator, "quorumDenom", QuorumDenominator)
+	if err := warpMsg.Signature.Verify(
+		context.Background(),
+		&warpMsg.UnsignedMessage,
+		predicateContext.SnowCtx.ValidatorState, // TODO: special case messages from the C-Chain
```
Comment (aaronbuchwald) [inline precompile/contracts/warp/config.go:132]: TODO

File: precompile/contracts/warp/config.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+		quorumNumerator,
+		QuorumDenominator,
+	); err != nil {
+		return 0, fmt.Errorf("warp signature verification failed: %w", err)
+	}
+
+	return msgGas, nil
+}
+
+// TODO: move to general package, cleanup, and test
```
Comment (aaronbuchwald) [inline precompile/contracts/warp/config.go:143]: TODO

File: precompile/contracts/warp/config.go (hunk 6 above)
Comment (aaronbuchwald) [reply precompile/contracts/warp/config.go:94]: TODO: add a README as well

File: precompile/contracts/warp/contract.go [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+	// Cost of fixed log size and writing log to database (Note: this calculation includes buffer room because a warp message
+	// is not stored in the trie and not stored permanently)
+	SendWarpMessageGasCost uint64 = params.LogGas + 4*params.LogTopicGas + contract.WriteGasCostPerSlot
+	// TODO: Gas cost per byte in the payload/log data of the produced warp message
+	SendWarpMessageGasCostPerByte uint64 = 1
+
+	GetVerifiedWarpMessageGasCost uint64 = 1 // TODO: base this off of the number of signers
+	GasCostPerWarpSigner          uint64 = 1 // TODO
+	GasCostPerWarpMessageBytes    uint64 = 1 // TODO
+)
```
Comment (aaronbuchwald) [inline precompile/contracts/warp/contract.go:35]: TODO: need to benchmark the warp precompile predicate and execution in order to calculate accurate gas costs.

File: precompile/contracts/warp/contract.go [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// PackGetVerifiedWarpMessage packs the include selector (first 4 func signature bytes).
+// This function is mostly used for tests.
```
Comment (aaronbuchwald) [inline precompile/contracts/warp/contract.go:95]: Clean up these comments and fix it in the precompile template

File: precompile/contracts/warp/contract.go [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+	predicateBytes, exists := accessibleState.GetStateDB().GetPredicateStorageSlots(ContractAddress)
+	// If there is no such value, return false to the caller.
+	// Note: decoding errors will return an error instead.
+	if !exists {
+		packedOutput, err := PackGetVerifiedWarpMessageOutput(GetVerifiedWarpMessageOutput{
+			Exists: false,
+		})
+		if err != nil {
+			return nil, remainingGas, err
+		}
```
Comment (aaronbuchwald) [inline precompile/contracts/warp/contract.go:123]: TODO: add Warn logs to unexpected ABI packing/unpacking errors.

File: precompile/contracts/warp/contract_test.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+			Caller:      callerAddr,
+			InputFn:     func(t *testing.T) []byte { return sendWarpMessageInput },
+			SuppliedGas: SendWarpMessageGasCost + uint64(len(sendWarpMessageInput[4:])*int(SendWarpMessageGasCostPerByte)),
+			ReadOnly:    false,
+			ExpectedRes: []byte{},
+			AfterHook: func(t *testing.T, state contract.StateDB) {
+				// TODO: check that the log was produced correctly
+			},
+		},
+		// TODO: add unit tests for getVerifiedWarpMessage
```
Comment (aaronbuchwald) [inline precompile/contracts/warp/contract_test.go:114]: TODO

File: warp/messages/codec.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package messages
```
Comment (aaronbuchwald) [inline warp/messages/codec.go:4]: Change to `payloads`

File: precompile/contracts/warp/contract.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+		return nil, remainingGas, err
+	}
+
+	// Add a log to be handled if this action is finalized.
+	accessibleState.GetStateDB().AddLog(
+		ContractAddress,
+		[]common.Hash{
+			WarpABI.Events["SendWarpMessage"].ID,
+			destinationChainID,
+			destinationAddress,
```
Comment (minghinmatthewlam) [inline precompile/contracts/warp/contract.go:235]: How did you decide on which topics to add to the log? And when receiving the log do we need to check that the ID/Addresses are the same? 

For example, parsing the data bytes back to warp unsigned message, then checking that topic[1] destination chain ID matches.

Comment (aaronbuchwald) [reply precompile/contracts/warp/contract.go:239]: Documented here: https://github.com/ava-labs/subnet-evm/blob/warp-e2e/precompile/contracts/warp/README.md#sendwarpmessage

File: plugin/evm/vm_warp_test.go [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/precompile/contracts/warp"
+	"github.com/ava-labs/subnet-evm/rpc"
+	warpTransaction "github.com/ava-labs/subnet-evm/warp/transaction"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/stretchr/testify/require"
+)
+
+// Test that sending a valid signed warp message results in successful delivery.
```
Comment (darioush) [inline plugin/evm/vm_warp_test.go:34]: could we update this comment since we don't deal with delivering the message to the other chain in subnet-evm?

File: core/state_transition.go [hunk 16]
```diff
@@ -163,12 +165,64 @@ func IntrinsicGas(data []byte, accessList types.AccessList, isContractCreation b
...
+
 	return gas, nil
 }
 
+func ApplyPredicateGas(rules params.Rules, accessTuple types.AccessTuple) (uint64, error) {
+	predicate, ok := rules.PredicatePrecompiles[accessTuple.Address]
+	if ok {
+		return predicate.PredicateGas(utils.HashSliceToBytes(accessTuple.StorageKeys))
+	}
+	proposerPredicates, ok := rules.ProposerPredicates[accessTuple.Address]
```
Comment (darioush) [inline core/state_transition.go:219]: nit: should be `proposerPredicate`
